### PR TITLE
feat: Listen on onion-service, dial onion services

### PR DESCRIPTION
**Pull request by binarybaron**
This PR adds support for listening on onion services (`/onion3/*:x/` addresses, where `*` is an onion address and `x` an arbitrary port)

It also add support for dialing them.

Closes https://github.com/umgefahren/libp2p-tor/issues/19

### CODE REVIEW

**Comment by binarybaron** [comment]
@umgefahren This works as a proof-of-concept. Can you take a look regarding the general direction this PR is going? 

```
# Starts listening on an onion service
cargo run --example ping-onion

# Connect to the onion service (adjust peer id and onion address)
cargo run --example ping-onion /onion3/nqxrqep7hqagkhs2x3l5vykzjhh3icov4yeukngwtype3hr7ys2inqid:9999/p2p/12D3KooWREFBSvaPqzdiCjaqC9jnySdQfecCUD9w8e1uztVJPLYF
```

**Comment by binarybaron** [comment]
@einliterflasche You can give this a review as well once it's ready.

**Comment by umgefahren** [comment]
I think you are going in the right direction. Although I would suggest you put this behind a feature gate

**Comment by binarybaron** [comment]
@umgefahren @einliterflasche This is ready for review

**Comment by binarybaron** [comment]
@umgefahren Do you have time to give this a review?

File: examples/ping-onion.rs [hunk 1]
```diff
@@ -104,18 +123,20 @@ async fn main() -> Result<(), Box<dyn Error>> {
...
-
     // Dial the peer identified by the multi-address given as the second
     // command-line argument, if any.
     if let Some(addr) = std::env::args().nth(1) {
         let remote: Multiaddr = addr.parse()?;
         swarm.dial(remote)?;
         println!("Dialed {addr}")
+    } else {
+        // TODO: We need to do this because otherwise the status of the onion service is gonna be [`Shutdown`]
+        // when we first poll it and then the swarm will not pull it again (?). I don't know why this is the case.
```
Comment (umgefahren) [inline examples/ping-onion.rs:134]: Please look into this. But this can also be fixed later.

**Review by umgefahren — APPROVED** [review]
Just some night pick. Otherwise LGTM :rocket:

**Pull request merged**